lib/db_connection.ex

```diff
@@ -146,42 +157,56 @@ defmodule DBConnection do
     {:ok, new_state :: any} | {:disconnect, Exception.t, new_state :: any}
 
   @doc """
-  Handle the beginning of a transaction. Return `{:ok, result, state}` to
-  continue, `{:error, exception, state}` to abort the transaction and
+  Handle the beginning of a transaction.
+
+  Return `{:ok, result, state}` to continue, `{:commit, state}` to notify caller
+  to commit open transaction before continuing, `{:rollback, state}` to notify
+  caller to rollback aborted transaction before continuing,
+  `{:error, exception, state}` to abort the transaction and
   continue or `{:disconnect, exception, state}` to abort the transaction
   and disconnect.
 
+  A callback implementation should only return `:commit` or `:rollback` if it
+  can determine the database's transaction status without side effect.
+
   This callback is called in the client process.
   """
   @callback handle_begin(opts :: Keyword.t, state :: any) ::
     {:ok, result, new_state :: any} |
+    {:commit | :rollback, new_state :: any} |
     {:error | :disconnect, Exception.t, new_state :: any}
 
   @doc """
   Handle committing a transaction. Return `{:ok, result, state}` on success and
-  to continue, `{:error, exception, state}` to abort the transaction and
-  continue or `{:disconnect, exception, state}` to abort the transaction
-  and disconnect.
+  to continue, `{:begin, state}` to open transaction before continuing,
+  `{:rollback, state}` to rollback aborted transaction,
+  `{:error, exception, state}` to abort the transaction and continue or
```

Comment:
In this case it probably needs rewording. The error will cause `{:error, exception}` to be returned by `rollback/2`. For example this might occur if the driver can't do a side effect free transaction status check.